### ISSUE TITLE
a zero-length file in this tar file never emits 'end'

### ISSUE THREAD

**Issue by lfdoherty** [npm/node-tar#148]
I reported this bug earlier as a pull request, but not well. This time I've created a repo that reproduces the bug, rather than get overambitious and try to create a test....

https://github.com/lfdoherty/zero-length-file-bug

**Comment by jchip** [comment]
I also ran into this problem.

With this code below, the `console.log` never execute for empty file and there's no other events.  The promise stays `pending`.

```js
const Path = require("path");
const Tar = require("tar");

const promise = Tar.x({
  file: Path.resolve("empty.tar"),
  C: process.cwd()
}).then(() => {
  console.log("tar extract done");
}).catch( err => {
  console.log("fail", err);
});

setTimeout(() => {
  console.log(promise);
}, 1000);
```

**Comment by isaacs** [comment]
This works for me with the code on master in this repo.

```js
const path = require('path')
const tar = require('./')
const fs = require('fs')
fs.writeFileSync('empty.tar', '')

const promise = tar.x({
  file: path.resolve('empty.tar'),
  C: process.cwd()
}).then(() => {
  console.log('tar extract done')
  clearTimeout(timer)
}).catch( err => {
  console.log('fail', err)
});

const timer = setTimeout(() => {
  console.log(promise)
}, 1000)
```

What version of tar are you using?

**Comment by jchip** [comment]
I am using tar 4.0.2.

I just created install a new copy and your sample doesn't work for me.

```
$ grep version node_modules/tar/package.json 
    "postversion": "npm publish",
    "preversion": "npm test",
  "version": "4.0.2"
$ uname -a 
Darwin L-SB8H1M2G8W-M.local 16.7.0 Darwin Kernel Version 16.7.0: Thu Jun 15 17:36:27 PDT 2017; root:xnu-3789.70.16~2/RELEASE_X86_64 x86_64
$ node -v
v6.11.3
$ npm -v
3.10.10
$ node tarx.js
Promise { <pending> }
$ 
```

**Comment by isaacs** [comment]
Ah!  I lied! I wasn't using master, I had the `fsm2` branch checked out.  This is fixed (somehow) by the changes in #155.  Most likely, it's a race condition in the file handling that's fixed by unifying on a single high-performance stream implementation in all cases.
